Unable to connect to websocket server using wss://

### Expected Behavior

When installing fresh Micronaut 4.4.0 application

and creating these 2 classes

```java
package com.example;

import io.micronaut.websocket.WebSocketSession;
import io.micronaut.websocket.annotation.ClientWebSocket;
import io.micronaut.websocket.annotation.OnMessage;
import io.micronaut.websocket.annotation.OnOpen;


@ClientWebSocket
public class TalosClientWebsocket implements AutoCloseable {

    private WebSocketSession session;

    @OnOpen
    public void onOpen(WebSocketSession session) {
        this.session = session;
    }

    @OnMessage
    public void onMessage(String message) {
      System.out.println(message);
    }

    @Override
    public void close() {
      session.close();
    }
}
```

and

```java
package com.example;


import io.micronaut.context.annotation.Context;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.websocket.WebSocketClient;
import jakarta.inject.Singleton;
import reactor.core.publisher.Flux;

@Context
@Singleton
public class TalosClientWebsocketConnection {


  public TalosClientWebsocketConnection(
    @Client("wss://websocket-echo.com") final WebSocketClient webSocketClient
    ) {
    Flux.from(webSocketClient.connect(TalosClientWebsocket.class, "/"))
      .blockFirst()
    ;
  }
}
```

I expect application to connect to websocket server.

### Actual Behaviour

The application fails with error

```
09:50:53.480 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [com.example.TalosClientWebsocketConnection] could not be loaded: Error instantiating bean of type  [com.example.TalosClientWebsocketConnection]

Message: Error opening WebSocket client session: Abnormal Closure
Path Taken: new TalosClientWebsocketConnection(WebSocketClient webSocketClient)
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [com.example.TalosClientWebsocketConnection] could not be loaded: Error instantiating bean of type  [com.example.TalosClientWebsocketConnection]

Caused by: io.micronaut.websocket.exceptions.WebSocketClientException: Error opening WebSocket client session: Abnormal Closure
	at io.micronaut.http.client.netty.websocket.NettyWebSocketClientHandler.handleCloseReason(NettyWebSocketClientHandler.java:262)
	at io.micronaut.http.netty.websocket.AbstractNettyWebSocketHandler.handlerRemoved(AbstractNettyWebSocketHandler.java:255)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:1138)
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
	at io.netty.channel.DefaultChannelPipeline.destroyDown(DefaultChannelPipeline.java:876)
	at io.netty.channel.DefaultChannelPipeline.destroyUp(DefaultChannelPipeline.java:844)
	at io.netty.channel.DefaultChannelPipeline.destroy(DefaultChannelPipeline.java:836)
	at io.netty.channel.DefaultChannelPipeline.access$700(DefaultChannelPipeline.java:46)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1392)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:215)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:195)
	at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:821)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:821)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

**it works correctly if I connect to non wss:// (e.g. ws://localhost:8080)**

These are the debug logs from the netty
```
09:53:26.410 [main] DEBUG i.n.u.i.l.InternalLoggerFactory - Using SLF4J as the default logging framework
09:53:26.411 [main] DEBUG io.netty.util.ResourceLeakDetector - -Dio.netty.leakDetection.level: simple
09:53:26.411 [main] DEBUG io.netty.util.ResourceLeakDetector - -Dio.netty.leakDetection.targetRecords: 4
09:53:26.478 [main] DEBUG i.n.c.MultithreadEventLoopGroup - -Dio.netty.eventLoopThreads: 20
09:53:26.482 [main] DEBUG i.n.u.concurrent.GlobalEventExecutor - -Dio.netty.globalEventExecutor.quietPeriodSeconds: 1
09:53:26.484 [main] DEBUG i.n.u.i.InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.initialSize: 1024
09:53:26.484 [main] DEBUG i.n.u.i.InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.maxSize: 4096
09:53:26.492 [main] DEBUG i.n.util.internal.PlatformDependent0 - -Dio.netty.noUnsafe: false
09:53:26.492 [main] DEBUG i.n.util.internal.PlatformDependent0 - Java version: 21
09:53:26.493 [main] DEBUG i.n.util.internal.PlatformDependent0 - sun.misc.Unsafe.theUnsafe: available
09:53:26.493 [main] DEBUG i.n.util.internal.PlatformDependent0 - sun.misc.Unsafe.copyMemory: available
09:53:26.493 [main] DEBUG i.n.util.internal.PlatformDependent0 - sun.misc.Unsafe.storeFence: available
09:53:26.493 [main] DEBUG i.n.util.internal.PlatformDependent0 - java.nio.Buffer.address: available
09:53:26.493 [main] DEBUG i.n.util.internal.PlatformDependent0 - direct buffer constructor: unavailable: Reflective setAccessible(true) disabled
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent0 - java.nio.Bits.unaligned: available, true
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent0 - jdk.internal.misc.Unsafe.allocateUninitializedArray(int): unavailable: class io.netty.util.internal.PlatformDependent0$7 cannot access class jdk.internal.misc.Unsafe (in module java.base) because module java.base does not export jdk.internal.misc to unnamed module @4232c52b
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent0 - java.nio.DirectByteBuffer.<init>(long, {int,long}): unavailable
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent - sun.misc.Unsafe: available
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent - -Dio.netty.tmpdir: /var/folders/4w/px_kfrk55b9dr69m1l0n0ym00000gn/T (java.io.tmpdir)
09:53:26.494 [main] DEBUG i.n.util.internal.PlatformDependent - -Dio.netty.bitMode: 64 (sun.arch.data.model)
09:53:26.495 [main] DEBUG i.n.util.internal.PlatformDependent - Platform: MacOS
09:53:26.495 [main] DEBUG i.n.util.internal.PlatformDependent - -Dio.netty.maxDirectMemory: -1 bytes
09:53:26.495 [main] DEBUG i.n.util.internal.PlatformDependent - -Dio.netty.uninitializedArrayAllocationThreshold: -1
09:53:26.495 [main] DEBUG io.netty.util.internal.CleanerJava9 - java.nio.ByteBuffer.cleaner(): available
09:53:26.495 [main] DEBUG i.n.util.internal.PlatformDependent - -Dio.netty.noPreferDirect: false
09:53:26.495 [main] DEBUG io.netty.channel.nio.NioEventLoop - -Dio.netty.noKeySetOptimization: false
09:53:26.495 [main] DEBUG io.netty.channel.nio.NioEventLoop - -Dio.netty.selectorAutoRebuildThreshold: 512
09:53:26.498 [main] DEBUG i.n.util.internal.PlatformDependent - org.jctools-core.MpscChunkedArrayQueue: available
09:53:26.515 [main] DEBUG io.netty.handler.ssl.OpenSsl - netty-tcnative not in the classpath; OpenSslEngine will be unavailable.
09:53:26.619 [main] DEBUG io.netty.handler.ssl.JdkSslContext - Default protocols (JDK): [TLSv1.3, TLSv1.2] 
09:53:26.619 [main] DEBUG io.netty.handler.ssl.JdkSslContext - Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384]
09:53:26.657 [main] DEBUG io.netty.channel.DefaultChannelId - -Dio.netty.processId: 56637 (auto-detected)
09:53:26.658 [main] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv4Stack: false
09:53:26.658 [main] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv6Addresses: false
09:53:26.661 [main] DEBUG i.netty.util.NetUtilInitializations - Loopback interface: lo0 (lo0, 0:0:0:0:0:0:0:1%lo0)
09:53:26.662 [main] DEBUG io.netty.util.NetUtil - Failed to get SOMAXCONN from sysctl and file /proc/sys/net/core/somaxconn. Default: 128
09:53:26.663 [main] DEBUG io.netty.channel.DefaultChannelId - -Dio.netty.machineId: f8:4d:89:ff:fe:93:80:c4 (auto-detected)
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.numHeapArenas: 20
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.numDirectArenas: 20
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.pageSize: 8192
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.maxOrder: 3
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.chunkSize: 65536
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.smallCacheSize: 256
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.normalCacheSize: 64
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.maxCachedBufferCapacity: 32768
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.cacheTrimInterval: 8192
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.cacheTrimIntervalMillis: 0
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.useCacheForAllThreads: false
09:53:26.675 [main] DEBUG i.n.buffer.PooledByteBufAllocator - -Dio.netty.allocator.maxCachedByteBuffersPerChunk: 1023
09:53:26.679 [main] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.allocator.type: pooled
09:53:26.679 [main] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.threadLocalDirectBufferSize: 0
09:53:26.679 [main] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.maxThreadLocalCharBufferSize: 16384
09:53:26.683 [main] DEBUG i.n.b.ChannelInitializerExtensions - -Dio.netty.bootstrap.extensions: null
09:53:26.703 [default-nioEventLoopGroup-1-1] DEBUG io.netty.buffer.AbstractByteBuf - -Dio.netty.buffer.checkAccessible: true
09:53:26.703 [default-nioEventLoopGroup-1-1] DEBUG io.netty.buffer.AbstractByteBuf - -Dio.netty.buffer.checkBounds: true
09:53:26.704 [default-nioEventLoopGroup-1-1] DEBUG i.n.util.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@4d315413
09:53:26.713 [default-nioEventLoopGroup-1-1] DEBUG i.n.h.c.compression.ZlibCodecFactory - -Dio.netty.noJdkZlibDecoder: false
09:53:26.713 [default-nioEventLoopGroup-1-1] DEBUG i.n.h.c.compression.ZlibCodecFactory - -Dio.netty.noJdkZlibEncoder: false
09:53:26.882 [default-nioEventLoopGroup-1-1] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.maxCapacityPerThread: 4096
09:53:26.882 [default-nioEventLoopGroup-1-1] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.ratio: 8
09:53:26.882 [default-nioEventLoopGroup-1-1] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.chunkSize: 32
09:53:26.882 [default-nioEventLoopGroup-1-1] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.blocking: false
09:53:26.882 [default-nioEventLoopGroup-1-1] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.batchFastThreadLocalOnly: true
09:53:26.887 [default-nioEventLoopGroup-1-1] DEBUG i.n.h.c.h.w.WebSocketClientHandshaker13 - WebSocket version 13 client handshake key: xWELLURO9uS4JZjTBIaXyw==, expected response: F5CLHu3C96X3B2xBs7T3e7vKb2M=
09:53:27.048 [default-nioEventLoopGroup-1-1] DEBUG io.netty.handler.ssl.SslHandler - [id: 0xd776b5e6, L:/192.168.100.43:57120 - R:websocket-echo.com/198.12.65.76:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_128_GCM_SHA256
```

### Steps To Reproduce

Install fresh micronaut 4.4.0 application

with these dependencies

```
```
dependencies {
    annotationProcessor("io.micronaut:micronaut-http-validation")
    annotationProcessor("io.micronaut.serde:micronaut-serde-processor")
    implementation("io.micronaut:micronaut-websocket")
    implementation("io.micronaut.reactor:micronaut-reactor-http-client")
    implementation("io.micronaut.serde:micronaut-serde-jackson")
    compileOnly("io.micronaut:micronaut-http-client")
    runtimeOnly("ch.qos.logback:logback-classic")
    testImplementation("io.micronaut:micronaut-http-client")
}
```


### Environment Information

- JDK 21
- MacOS

### Example Application

https://github.com/iNviNho/micronaut-wss-issue

### Version

4.4.0

--- a/http-client-core/src/main/java/io/micronaut/http/client/HttpVersionSelection.java
+++ b/http-client-core/src/main/java/io/micronaut/http/client/HttpVersionSelection.java
@@ -63,6 +63,12 @@ public final class HttpVersionSelection {
         true
     );
 
+    private static final HttpVersionSelection WEBSOCKET_1 = new HttpVersionSelection(
+        HttpVersionSelection.PlaintextMode.HTTP_1,
+        true,
+        new String[]{HttpVersionSelection.ALPN_HTTP_1},
+        false);
+
     private final PlaintextMode plaintextMode;
     private final boolean alpn;
     private final String[] alpnSupportedProtocols;
@@ -98,6 +104,17 @@ public final class HttpVersionSelection {
             default:
                 throw new IllegalArgumentException("HTTP version " + httpVersion + " not supported here");
         }
+    }
+
+    /**
+     * Get the {@link HttpVersionSelection} to be used for a WebSocket connection, which will enable
+     * ALPN but constrain the mode to HTTP 1.1.
+     *
+     * @return The version selection for WebSocket
+     */
+    @NonNull
+    public static HttpVersionSelection forWebsocket() {
+        return WEBSOCKET_1;
     }
 
     /**

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -22,6 +22,7 @@ import io.micronaut.core.propagation.PropagatedContext;
 import io.micronaut.core.reflect.InstantiationUtils;
 import io.micronaut.core.util.StringUtils;
 import io.micronaut.core.util.SupplierUtil;
+import io.micronaut.http.HttpVersion;
 import io.micronaut.http.client.HttpClientConfiguration;
 import io.micronaut.http.client.HttpVersionSelection;
 import io.micronaut.http.client.exceptions.HttpClientException;
@@ -143,6 +144,7 @@ public class ConnectionManager {
     private Bootstrap udpBootstrap;
     private final HttpClientConfiguration configuration;
     private volatile SslContext sslContext;
+    private volatile SslContext websocketSslContext;
     private volatile /* QuicSslContext */ Object http3SslContext;
     private final NettyClientCustomizer clientCustomizer;
     private final String informationalServiceId;
@@ -164,6 +166,7 @@ public class ConnectionManager {
         this.udpBootstrap = from.udpBootstrap;
         this.configuration = from.configuration;
         this.sslContext = from.sslContext;
+        this.websocketSslContext = from.websocketSslContext;
         this.http3SslContext = from.http3SslContext;
         this.clientCustomizer = from.clientCustomizer;
         this.informationalServiceId = from.informationalServiceId;
@@ -209,10 +212,13 @@ public class ConnectionManager {
 
     final void refresh() {
         SslContext oldSslContext = sslContext;
+        SslContext oldWebsocketSslContext = websocketSslContext;
         if (configuration.getSslConfiguration().isEnabled()) {
             sslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), httpVersion);
+            websocketSslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forLegacyVersion(HttpVersion.HTTP_1_1));
         } else {
             sslContext = null;
+            websocketSslContext = null;
         }
         if (httpVersion.isHttp3()) {
             http3SslContext = nettyClientSslBuilder.buildHttp3(configuration.getSslConfiguration());
@@ -224,6 +230,7 @@ public class ConnectionManager {
             pool.forEachConnection(c -> ((Pool.ConnectionHolder) c).windDownConnection());
         }
         ReferenceCountUtil.release(oldSslContext);
+        ReferenceCountUtil.release(oldWebsocketSslContext);
     }
 
     /**
@@ -433,6 +440,26 @@ public class ConnectionManager {
     }
 
     /**
+     * Builds an {@link SslContext} for the given WebSocket URI if necessary.
+     *
+     * @return The {@link SslContext} instance
+     */
+    @Nullable
+    private SslContext buildWebsocketSslContext(DefaultHttpClient.RequestKey requestKey) {
+        final SslContext sslCtx;
+        if (requestKey.isSecure()) {
+            sslCtx = websocketSslContext;
+            //Allow wss requests to be sent if SSL is disabled but a proxy is present
+            if (sslCtx == null && !configuration.getProxyAddress().isPresent()) {
+                throw decorate(new HttpClientException("Cannot send WSS request. SSL is disabled"));
+            }
+        } else {
+            sslCtx = null;
+        }
+        return sslCtx;
+    }
+
+    /**
      * Connect to a remote websocket. The given {@link ChannelHandler} is added to the pipeline
      * when the handshakes complete.
      *
@@ -448,7 +475,7 @@ public class ConnectionManager {
             protected void initChannel(@NonNull Channel ch) {
                 addLogHandler(ch);
 
-                SslContext sslContext = buildSslContext(requestKey);
+                SslContext sslContext = buildWebsocketSslContext(requestKey);
                 if (sslContext != null) {
                     ch.pipeline().addLast(configureSslHandler(sslContext.newHandler(ch.alloc(), requestKey.getHost(), requestKey.getPort())));
                 }

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -215,7 +215,7 @@ public class ConnectionManager {
         SslContext oldWebsocketSslContext = websocketSslContext;
         if (configuration.getSslConfiguration().isEnabled()) {
             sslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), httpVersion);
-            websocketSslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forLegacyVersion(HttpVersion.HTTP_1_1));
+            websocketSslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
         } else {
             sslContext = null;
             websocketSslContext = null;
@@ -376,7 +376,9 @@ public class ConnectionManager {
             }
         }
         ReferenceCountUtil.release(sslContext);
+        ReferenceCountUtil.release(websocketSslContext);
         sslContext = null;
+        websocketSslContext = null;
     }
 
     /**

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -22,7 +22,6 @@ import io.micronaut.core.propagation.PropagatedContext;
 import io.micronaut.core.reflect.InstantiationUtils;
 import io.micronaut.core.util.StringUtils;
 import io.micronaut.core.util.SupplierUtil;
-import io.micronaut.http.HttpVersion;
 import io.micronaut.http.client.HttpClientConfiguration;
 import io.micronaut.http.client.HttpVersionSelection;
 import io.micronaut.http.client.exceptions.HttpClientException;
@@ -144,7 +143,6 @@ public class ConnectionManager {
     private Bootstrap udpBootstrap;
     private final HttpClientConfiguration configuration;
     private volatile SslContext sslContext;
-    private volatile SslContext websocketSslContext;
     private volatile /* QuicSslContext */ Object http3SslContext;
     private final NettyClientCustomizer clientCustomizer;
     private final String informationalServiceId;
@@ -166,7 +164,6 @@ public class ConnectionManager {
         this.udpBootstrap = from.udpBootstrap;
         this.configuration = from.configuration;
         this.sslContext = from.sslContext;
-        this.websocketSslContext = from.websocketSslContext;
         this.http3SslContext = from.http3SslContext;
         this.clientCustomizer = from.clientCustomizer;
         this.informationalServiceId = from.informationalServiceId;
@@ -212,13 +209,10 @@ public class ConnectionManager {
 
     final void refresh() {
         SslContext oldSslContext = sslContext;
-        SslContext oldWebsocketSslContext = websocketSslContext;
         if (configuration.getSslConfiguration().isEnabled()) {
             sslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), httpVersion);
-            websocketSslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
         } else {
             sslContext = null;
-            websocketSslContext = null;
         }
         if (httpVersion.isHttp3()) {
             http3SslContext = nettyClientSslBuilder.buildHttp3(configuration.getSslConfiguration());
@@ -230,7 +224,6 @@ public class ConnectionManager {
             pool.forEachConnection(c -> ((Pool.ConnectionHolder) c).windDownConnection());
         }
         ReferenceCountUtil.release(oldSslContext);
-        ReferenceCountUtil.release(oldWebsocketSslContext);
     }
 
     /**
@@ -376,9 +369,7 @@ public class ConnectionManager {
             }
         }
         ReferenceCountUtil.release(sslContext);
-        ReferenceCountUtil.release(websocketSslContext);
         sslContext = null;
-        websocketSslContext = null;
     }
 
     /**
@@ -448,17 +439,17 @@ public class ConnectionManager {
      */
     @Nullable
     private SslContext buildWebsocketSslContext(DefaultHttpClient.RequestKey requestKey) {
-        final SslContext sslCtx;
         if (requestKey.isSecure()) {
-            sslCtx = websocketSslContext;
-            //Allow wss requests to be sent if SSL is disabled but a proxy is present
-            if (sslCtx == null && !configuration.getProxyAddress().isPresent()) {
+            if (configuration.getSslConfiguration().isEnabled()) {
+                return nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
+            } else if (configuration.getProxyAddress().isEmpty()){
                 throw decorate(new HttpClientException("Cannot send WSS request. SSL is disabled"));
+            } else {
+                return null;
             }
         } else {
-            sslCtx = null;
+            return null;
         }
-        return sslCtx;
     }
 
     /**

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -144,6 +144,7 @@ public class ConnectionManager {
     private final HttpClientConfiguration configuration;
     private volatile SslContext sslContext;
     private volatile /* QuicSslContext */ Object http3SslContext;
+    private volatile SslContext websocketSslContext;
     private final NettyClientCustomizer clientCustomizer;
     private final String informationalServiceId;
 
@@ -165,6 +166,7 @@ public class ConnectionManager {
         this.configuration = from.configuration;
         this.sslContext = from.sslContext;
         this.http3SslContext = from.http3SslContext;
+        this.websocketSslContext = from.websocketSslContext;
         this.clientCustomizer = from.clientCustomizer;
         this.informationalServiceId = from.informationalServiceId;
         this.nettyClientSslBuilder = from.nettyClientSslBuilder;
@@ -209,6 +211,8 @@ public class ConnectionManager {
 
     final void refresh() {
         SslContext oldSslContext = sslContext;
+        SslContext oldWebsocketSslContext = websocketSslContext;
+        websocketSslContext = null;
         if (configuration.getSslConfiguration().isEnabled()) {
             sslContext = nettyClientSslBuilder.build(configuration.getSslConfiguration(), httpVersion);
         } else {
@@ -224,6 +228,7 @@ public class ConnectionManager {
             pool.forEachConnection(c -> ((Pool.ConnectionHolder) c).windDownConnection());
         }
         ReferenceCountUtil.release(oldSslContext);
+        ReferenceCountUtil.release(oldWebsocketSslContext);
     }
 
     /**
@@ -369,7 +374,9 @@ public class ConnectionManager {
             }
         }
         ReferenceCountUtil.release(sslContext);
+        ReferenceCountUtil.release(websocketSslContext);
         sslContext = null;
+        websocketSslContext = null;
     }
 
     /**
@@ -439,17 +446,22 @@ public class ConnectionManager {
      */
     @Nullable
     private SslContext buildWebsocketSslContext(DefaultHttpClient.RequestKey requestKey) {
+        SslContext sslCtx = websocketSslContext;
         if (requestKey.isSecure()) {
             if (configuration.getSslConfiguration().isEnabled()) {
-                return nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
+                if (sslCtx == null) {
+                    synchronized (this) {
+                        sslCtx = websocketSslContext;
+                        if (sslCtx == null) {
+                            sslCtx = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
+                        }
+                    }
+                }
             } else if (configuration.getProxyAddress().isEmpty()){
                 throw decorate(new HttpClientException("Cannot send WSS request. SSL is disabled"));
-            } else {
-                return null;
             }
-        } else {
-            return null;
         }
+        return sslCtx;
     }
 
     /**

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -454,6 +454,7 @@ public class ConnectionManager {
                         sslCtx = websocketSslContext;
                         if (sslCtx == null) {
                             sslCtx = nettyClientSslBuilder.build(configuration.getSslConfiguration(), HttpVersionSelection.forWebsocket());
+                            websocketSslContext = sslCtx;
                         }
                     }
                 }

--- a/http-client/src/test/groovy/io/micronaut/http/client/websocket/ClientWebsocketSpec.groovy
+++ b/http-client/src/test/groovy/io/micronaut/http/client/websocket/ClientWebsocketSpec.groovy
@@ -11,6 +11,7 @@ import io.micronaut.websocket.exceptions.WebSocketClientException
 import jakarta.inject.Inject
 import jakarta.inject.Singleton
 import reactor.core.publisher.Mono
+import spock.lang.Issue
 import spock.lang.Specification
 
 import java.util.concurrent.ExecutionException
@@ -31,6 +32,47 @@ class ClientWebsocketSpec extends Specification {
 
         registry.clientBeans.size() == 1
         !registry.clientBeans[0].opened
+        !registry.clientBeans[0].autoClosed
+        !registry.clientBeans[0].onClosed
+
+        cleanup:
+        client.close()
+    }
+
+    @Issue("https://github.com/micronaut-projects/micronaut-core/issues/10744")
+    void 'websocket bean can connect to echo server over SSL with wss scheme'() {
+        given:
+        def ctx = ApplicationContext.run(['spec.name': 'ClientWebsocketSpec'])
+        def client = ctx.getBean(WebSocketClient)
+        def registry = ctx.getBean(ClientBeanRegistry)
+        def mono = Mono.from(client.connect(ClientBean.class, 'wss://websocket-echo.com'))
+
+        when:
+        mono.toFuture().get()
+
+        then:
+        registry.clientBeans.size() == 1
+        registry.clientBeans[0].opened
+        !registry.clientBeans[0].autoClosed
+        !registry.clientBeans[0].onClosed
+
+        cleanup:
+        client.close()
+    }
+
+    void 'websocket bean can connect to echo server over SSL with https scheme'() {
+        given:
+        def ctx = ApplicationContext.run(['spec.name': 'ClientWebsocketSpec'])//, "micronaut.http.client.alpn-modes":"http/1.1"])
+        def client = ctx.getBean(WebSocketClient)
+        def registry = ctx.getBean(ClientBeanRegistry)
+        def mono = Mono.from(client.connect(ClientBean.class, 'https://websocket-echo.com'))
+
+        when:
+        mono.toFuture().get()
+
+        then:
+        registry.clientBeans.size() == 1
+        registry.clientBeans[0].opened
         !registry.clientBeans[0].autoClosed
         !registry.clientBeans[0].onClosed
 

